proxy/forward_auth.go

```diff
@@ -119,21 +119,27 @@ func (p *Proxy) Verify(verifyOnly bool) http.Handler {
 		}
 
 		original := p.getOriginalRequest(r, uri)
-		authorized, err := p.isAuthorized(w, original)
+		isAuthorized, needsAuthentication, err := p.isAuthorized(w, original)
 		if err != nil {
 			return httputil.NewError(http.StatusBadRequest, err)
 		}
 
-		if authorized {
+		if isAuthorized {
 			w.Header().Set("Content-Type", "text/plain; charset=utf-8")
 			w.WriteHeader(http.StatusOK)
 			fmt.Fprintf(w, "Access to %s is allowed.", uri.Host)
 			return nil
 		}
 
+		//if unauthorized (for example, session is invalid), clear the session here too
+		if needsAuthentication {
+			p.sessionStore.ClearSession(w, r)
+		}
+
+		//if we have a session, and we didn't return unauthorized, return access denied
 		_, err = sessions.FromContext(r.Context())
 		hasSession := err == nil
-		if hasSession {
+		if hasSession && !needsAuthentication {
```

Comment:
This sounds not right to me, must we also change the `verifiedOnly` condition below?

Reply:
We shouldn't need to change the verifyOnly condition.  Basically, it's a flow through.  If it reaches the point after the modified block, it requires re-authentication.  The verifyOnly flag just allows it to be broken into two steps versus completing within this function (for NGINX - see the registration).